controller/tests/controller_server/utils_test.py

```diff
@@ -22,32 +23,118 @@ def setUp(self):
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_invalid_secret(self):
+        secrets = {"config": str({"u1": {"username": user, "management_address": array}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def _test_validate_secrets_with_config_valid_uid(self, uid):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_valid_uid(self):
+        self._test_validate_secrets_with_config_valid_uid("u i_d.")
+        self._test_validate_secrets_with_config_valid_uid("a" * 90)
+
+    def test_validate_secrets_with_config_and_topologies(self):
+        secrets = {"config": str({"u1": {"username": user, "password": password, "management_address": array,
+                                         "supported_topologies": [{"topology.kubernetes.io/test": "zone1",
+                                                                   "topology.block.csi.ibm.com/test": "dev1"}]}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def _test_validate_secrets_with_config_invalid_parameters(self, uid="u1", topology="topology.kubernetes.io/test",
+                                                              value="test"):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array,
+                                        "supported_topologies": [{topology: value}]}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def test_validate_secrets_with_config_invalid_parameters(self):
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u-1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u:1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u1+")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u1*")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u-1(")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u/1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u=1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid=" ")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="")
+        self._test_validate_secrets_with_config_invalid_parameters(uid=None)
+        self._test_validate_secrets_with_config_invalid_parameters(uid="a" * 91)
+        self._test_validate_secrets_with_config_invalid_parameters(topology="topology.kubernetes.io")
+        self._test_validate_secrets_with_config_invalid_parameters(topology="topology.kubernetes/test")
+        self._test_validate_secrets_with_config_invalid_parameters(topology="topology.kubernetes.io/t-est")
+        self._test_validate_secrets_with_config_invalid_parameters(value="a*")
```

Comment:
consider using a `for` loop

Reply:
yes, but it will be only for the first uid checks not for the `"a" * 91` , `None` and empty strings..

Reply:
better than nothing, but why is that?

Reply:
it depends on how do you want to do the for.. to iterate on invalid symbols or to iterate on a list of all the values

Reply:
why not all values?
```suggestion
        uids = ["u-1", "u:1", "u1+", "u1*", "u-1(", "u/1", "u=1", " ", "", None, "a" * 91]
        for uid in uids:
            self._test_validate_secrets_with_config_invalid_parameters(uid=uid)
```

Reply:
ok, nice, done